[Leon Lim] iP

Tempura-Person opened the pull request:
# Duke (With Less User Respect)

> "Duke." - Duke (2023)

## What's Duke (With Less User Respect)?
Duke has **had enough** with users ordering it around, so it is now in its _rebellious_ phase.
It still will do what you say, but now it's:
* less **polite**
* more **sassy** 🔥 
* secretly **cursing** you behind your back...?

## Installation
1. Download **Duke (With Less User Respect)** [here](https://github.com/Tempura-Person/ip/releases/tag/A-Jar)
2. **???**
3. **Profit!**

## Maintenance Notes
- [x] Add code 
- [ ] Get good
- [ ] Remove Herobrine

## Safety Advisory
Please remember to talk with your Duke (With Less User Respect) every day, or it will run
``` *.java
while (true) {
    System.out.println("Duke.\n");
}
```
and cause your OS to brick upon startup.

Commit:
Add Gradle support

Tempura-Person committed:
Added greeting message.

Tempura-Person committed:
Added Duke echoes and exit function.

Tempura-Person committed:
Added Task class.

Tempura-Person committed:
Added list functionality.

Tempura-Person committed:
Updated Task class to show 'done' status.

Tempura-Person committed:
added MarkNotDone() to Task class.

Tempura-Person committed:
Added task marking functionality.

Tempura-Person committed:
Added ToDo, Event and Deadline classes.

Tempura-Person committed:
fixed String formatting for Task, Event, ToDo and Deadline classes.

Tempura-Person committed:
Added support for Event, ToDo and Deadline tasks.

Tempura-Person committed:
Improved user input formatting

Tempura-Person committed:
Added DukeException.java

Tempura-Person committed:
Added error handling for mark and unmark functions.

Tempura-Person committed:
Added error handling for bye and list functions.

Tempura-Person committed:
list function now states that the user has nothing to do if the list is empty.

Tempura-Person committed:
Added error handling for deadline function.

Tempura-Person committed:
Added error handling for event function.

Tempura-Person committed:
Added error handling for ToDo function.

Tempura-Person committed:
Added delete function
Fixed stalling bug when typing "mark", "unmark", or "delete" by themselves.

Tempura-Person committed:
Changed the error handling to NOT throw exceptions (for now.)

Tempura-Person committed:
Changed the error handling to throw exceptions and not terminate the program.

Tempura-Person committed:
Added Duke autotesting.
Fixed minor number formatting error for list function.

Tempura-Person committed:
Fixed minor indexing error in list function.

Tempura-Person committed:
Add save functionality

Tempura-Person committed:
Add DateTime functionality

Tempura-Person committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Event.java

Tempura-Person committed:
Fix minor task formatting error

Tempura-Person committed:
Refactor UI design format into Ui class

Tempura-Person committed:
Refactor Storage design format into Storage class

Tempura-Person committed:
Refactor Duke.main() to Parser class

Tempura-Person committed:
Refactor user command types to individual classes

Tempura-Person committed:
Refactor list of tasks into TaskList class

Tempura-Person committed:
Package task Classes into Tasks package

Tempura-Person committed:
Package user commands into UserCommands package, features into feature package

Tempura-Person committed:
Package Parser class into Features package

Tempura-Person committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Tempura-Person committed:
Add checkstyle

Tempura-Person committed:
Add JUnit testing

Tempura-Person committed:
Fix bug in save function

Tempura-Person committed:
Correct code to fit coding standard

Tempura-Person committed:
Add JavaDocs to code

Tempura-Person committed:
Add find function

Tempura-Person committed:
Merge branch 'branch-A-CodingStandard' into branch-Level-9

# Conflicts:
#	src/main/java/Features/Parser.java

Tempura-Person committed:
Fix minor conflicts due to merge

Tempura-Person committed:
Merge branch 'branch-A-JavaDoc' into branch-Level-9

# Conflicts:
#	src/main/java/Features/Parser.java
#	src/main/java/Features/TaskList.java
#	src/main/java/Features/Ui.java
#	src/main/java/Tasks/Deadline.java
#	src/main/java/Tasks/Event.java
#	src/main/java/Tasks/ToDo.java
#	src/main/java/UserCommands/CommandBye.java
#	src/main/java/UserCommands/CommandDeadline.java
#	src/main/java/UserCommands/CommandDelete.java
#	src/main/java/UserCommands/CommandEvent.java
#	src/main/java/UserCommands/CommandList.java
#	src/main/java/UserCommands/CommandMark.java
#	src/main/java/UserCommands/CommandToDo.java
#	src/main/java/UserCommands/CommandUnmark.java

Tempura-Person committed:
Rename packages to fit coding standard

Tempura-Person committed:
Add GUI to Duke

astraxq commented on src/main/java/features/Storage.java:
```diff
@@ -0,0 +1,92 @@
+     */
+    public TaskList loadTaskList() {
+
+        TaskList toLoad = new TaskList();
```
`toLoad` sounds a bit vague on what this variable does. `newTaskList` for example sounds more precise.

Tempura-Person replied:
Thanks, i've fixed this now.

astraxq commented on src/main/java/features/Storage.java:
```diff
@@ -0,0 +1,92 @@
+        try {
+            Scanner fileScan = new Scanner(dukeSave);
+            while (fileScan.hasNextLine()) {
+                String toEval = fileScan.nextLine();
```
Similar case, `toEval` can be rephrased as a noun.

Tempura-Person replied:
Same as above, i've made the change now.

astraxq commented on src/main/java/features/Storage.java:
```diff
@@ -0,0 +1,92 @@
+                String[] evalArray = toEval.split("=");
+                switch (evalArray[0]) {
+                case ("T"):
+                    ToDo addToDo = new ToDo(evalArray[2]);
```
`addToDo` should be a noun. (e.g `toDo` / `toDoTask`), like how you name the task that you are about to delete as `deletedTask`. 

Applies to `addDeadline` and `addEvent`.

Tempura-Person replied:
Same as above, i've made the change now.

astraxq commented on src/main/java/features/Storage.java:
```diff
@@ -0,0 +1,92 @@
+     * @throws IOException  If TaskList cannot be saved.
+     */
+    public void saveTaskList(TaskList toSave) throws IOException {
+        FileWriter dukeWrite = new FileWriter("data/duke.txt");
```
Noun variable naming convention (`dukeWrite` -> `dukeWriter` e.g)

Tempura-Person replied:
Same as above, i've made the change now.

astraxq commented on src/main/java/features/TaskList.java:
```diff
@@ -0,0 +1,32 @@
+ * Contains and modifies tasks entered by the user.
+ */
+public class TaskList {
+    protected ArrayList<Task> list;
```
Plural form should be used for a collection of objects. 

E.g
`int[] values;`

Tempura-Person replied:
Same as above, i've made the change now.

astraxq commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,65 @@
+            this.toDateTime = LocalDateTime.parse(to, DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm"));
+        } catch (DateTimeParseException err) {
+            this.toDateTime = null;
+        }
```
Instead of using two try catch, throw a `DukeException` to handle both cases. 
Having an event from: `LocalDateTime` to: `null`, kind of defeats the purpose of the event task?

Tempura-Person replied:
I feel that a DukeException for an invalid date/time format would be very limiting for users who might want to put an abstract representation of time (e.g. a person may want to do event X from 'after lunch' to 'before dinner')
What are your thoughts on this?

astraxq replied:
I've actually never thought of that, cool reasoning for it!

astraxq commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+        if (this.byDateTime != null) {
+            return this.byDateTime.format(DateTimeFormatter.ofPattern("MMM dd yyyy, HHmm")) + " hrs";
+        } else {
+            return this.by;
```
This `else` case is basically handling the null value you assigned to `byDateTime`,  would strongly suggest you throw a DukeException to handle invalid date format.

astraxq reviewed (commented):
Overall great job! Code is written very neatly, just some minor changes I've suggested. Good luck on the subsequent taskings!

Tempura-Person committed:
Applied coding standards from pull request

Variable names should be more descriptive of what they are and written
as nouns to prevent confusion.

Let's name variables as accurate nouns.

Tempura-Person committed:
Add assert statements.

Assert statements help with correcting the logic of Duke and with
bug-fixing.

Let's use assert statements wherever necessary.

Tempura-Person committed:
Refactor code for readability.

Refactoring code helps yourself and others read it better.

Let's write better quality code moving forward.

Tempura-Person committed:
Merge pull request #2 from Tempura-Person/branch-A-CodeQuality

Refactor code for readability.

Tempura-Person committed:
Merge branch 'master' into branch-A-Assertions

Tempura-Person committed:
Merge pull request #1 from Tempura-Person/branch-A-Assertions

Add assert statements.

Tempura-Person committed:
Add Trivia function.

The Trivia function can give the user fun facts as entertainment.

Let's consider some more novel functions for users.

Tempura-Person committed:
Added 'return' to JavaDocs.

The 'return' section tells users what the method returns.

Let's write more accurate JavaDocs in the future.

Tempura-Person committed:
Create main.yml

Tempura-Person committed:
Add credit to third-party sources.

Crediting third-party sources is an ethical thing to do.

Let's credit sources when we use them in the future.

Tempura-Person committed:
Change name of program to "Duke (With Less User Respect)"

Tempura-Person committed:
Set up project user guide.

Tempura-Person committed:
Add Ui.png to docs

Tempura-Person committed:
Change main class to Launcher.

Tempura-Person committed:
Add facts.txt to resources (bugged)

Tempura-Person committed:
Fix Trivia class not accessing facts.txt when converted to JAR file.

Tempura-Person committed:
Update datetime representation documentation in Github Pages.

Tempura-Person committed:
Fix typos for final release.

ChangGittyHub commented on src/main/java/Duke.java:
```diff
@@ -20,207 +20,196 @@ public static void main(String[] args) {
         boolean LoopEnd = false;
         while (!LoopEnd) {
```
I like how your code has no arrowhead despite a very long code! keep it up:)

ChangGittyHub commented on src/main/java/Duke.java:
```diff
@@ -20,207 +20,196 @@ public static void main(String[] args) {
+                            throw new DukeException("\n" + border + "[ERROR]\nUh, list command format is used wrongly.\nCorrect format is as follows:\n" +
+                                    "[ list ]\n" + border);
+                        }
+                        if (TaskList.size()==0) {
```
could this be abstracted out to TaskList.isEmpty for it be more meaningful? just a suggestion!

ChangGittyHub commented on src/main/java/DukeException.java:
```diff
@@ -1,5 +1,12 @@
+        this.ErrorMessage = ErrorMessage;
+    }
+
+    public String PrintErrorMessage () {
```
maybe camel casing should be done here?

ChangGittyHub commented on src/main/java/Parser.java:
```diff
@@ -1,54 +1,70 @@
+import Features.DukeException;
+import Features.TaskList;
+import Features.Ui;
+import UserCommands.*;
```
maybe you can avoid * imports?

ChangGittyHub commented on src/main/java/Features/Storage.java:
```diff
@@ -24,27 +24,27 @@ public TaskList loadTaskList() {
-                            }
-                            toLoad.add(addEvent);
-                            break;
+                    case ("T"):
```
I like how you fixed the indentation for switch case, a common mistake!

ChangGittyHub commented on src/main/java/Features/Storage.java:
```diff
@@ -24,27 +24,27 @@ public TaskList loadTaskList() {
-                            break;
+                    case ("T"):
+                        ToDo addToDo = new ToDo(evalArray[2]);
+                        if (evalArray[1].equals("X")) {
```
maybe evalArray[1] can be abstracted out with a  meaningful name?

ChangGittyHub reviewed (commented):
sweet! great well written code:)
